openstf/model_selection/model_selection.py

```diff
@@ -1,87 +1,254 @@
+MIN_QUANTILE = 0.05
+
+
+def find_min_max_peaks(data: pd.DataFrame) -> (List[float], List[float]):
+    """
+    Retrieve minimum and maximum quantiles for a day (start till end)
+
+    Args:
+        data (pandas.DataFrame): Clean data with features
+
+    Returns:
+        min_peaks (list:float): Minimum peaks of a day
+        max_peaks (list:float): Maximum peaks of a day
+
+    """
+
+    try:
+        min_peaks = data["days"][:][data["load"] < data["load"].quantile(MIN_QUANTILE)]
+    except ValueError:
+        min_peaks = np.nan
+
+    try:
+        max_peaks = data["days"][:][data["load"] > data["load"].quantile(MAX_QUANTILE)]
+    except ValueError:
+        max_peaks = np.nan
+
+    return min_peaks, max_peaks
+
+
+def find_corresponding_min_max_days(data: pd.DataFrame) -> np.array:
+    """
+    Checks if there is a peak present, there is at least one peak (minimum or maximum)
+
+    Args:
+        data (pandas.DataFrame): Clean data with features
+    Returns:
+        present_peaks (np.array): Array filled with the days with peaks
+
+    """
+
+    min_peaks = find_min_max_peaks(data)[0].to_numpy()
+    max_peaks = find_min_max_peaks(data)[1].to_numpy()
+
+    present_peaks = np.unique(np.append(min_peaks, max_peaks))
+
+    return present_peaks
+
+
+def sample_indices_train_val(data: pd.DataFrame, peaks: List[int]) -> np.array:
+    """
+    Sample indices of given period length assuming the peaks are evenly spreaded.
+
+    Args:
+        data (pandas.DataFrame): Clean data with features
+        peaks (list:int): List of selected peaks to sample the indices from
+
+    Returns:
+        np.array: Sorted list with the indices corresponding to the peak
+
+    """
+
+    sampled = set()
+
+    for peak in peaks:
+        sampled |= set(data[data["days"] == peak].idx_num.to_numpy())
+    return np.sort(list(sampled))
+
+
+def random_sample(list_sample: List[int], k: int) -> List[int]:
+    """
+    Random sampling of numbers out of a list
+    (implemented due to security sonar cloud not accepting the random built-in functions)
+
+    Args:
+        list_sample (list:int): List with numbers to sample from
+        k (int): Number of wanted samples
+
+    Returns:
+        (list:int): Sorted list with the random samples
+
+    """
+
+    random_list = []
+    for _ in range(k):
+        element_random = secrets.choice(list_sample)
+        list_sample.remove(element_random)
+        random_list.append(element_random)
+    return random_list
 
 
 def split_data_train_validation_test(
-    data: pd.DataFrame,
-    test_fraction: float = 0.0,
+    data_: pd.DataFrame,
+    test_fraction: float = 0.1,
     validation_fraction: float = 0.15,
-    backtest: bool = False,
-    period_sampling: bool = True,
-    period_timedelta: timedelta = timedelta(days=2),
-) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
-    """Split input data into train, test and validation set.
+    back_test: bool = False,
+    stratification: bool = True,
```

Comment:
Perhaps renaming `stratification` to `stratify_min_max` makes it even more clear that this is not just the 'regular' stratification